### ISSUE TITLE
Fix Coverity Defects

### ISSUE THREAD

**Issue by anonimal** [monero-project/kovri#263]
**By submitting this issue, I confirm the following:**
- I have read and understood the [contributor guide](https://github.com/monero-project/kovri/blob/master/doc/CONTRIBUTING.md).
- I have checked that the issue I am reporting can be replicated or that the feature I am suggesting is not present.
- I have checked opened or recently closed [pull requests](https://github.com/monero-project/kovri/pulls) for existing solutions/implementations to my issue/suggestion.

_Place an X inside the bracket to confirm_
- [X] I confirm.

---

Coverity has proven itself to be problematic:
1. #209 
2. Less than 2 weeks ago, they' gave us a passing OK of 0 defects - which was false. For no inexplicable reason, they [now report that we have 39 defects](https://scan.coverity.com/projects/monero-project-kovri?tab=overview) and, from what I've read from a small snippet of email they sent me, are almost all in old inherited code (all being the snippet they sent me). This means that the **same code** that produced 0 defects less than two weeks ago is now defect worthy.
3. The website won't allow me to view the defects despite having correct permissions (see 1.)

Of the 39 existing defects, I have a feeling that a fair amount are either upstream (components for metrics need fine-tuning) and/or false-positives. I cannot confirm this yet until @fluffypony can take screenshots of the defects or if Coverity emails me _all_ the defects or if Coverity magically fixes their website.

When I have a better idea of where we are, I'll relabel this issue as needed.

**Comment by anonimal** [comment]
Coverity's website is still not allowing me to view defects. An email has been sent and CC'd to @fluffypony.

**Comment by anonimal** [comment]
They've fixed permissions and I've skimmed over the defects. The components were able to focus only on our codebase.

```
&anonimal | Well, 37 of the 39 are not my code. With the remaining 2, 1 is a questionable false-positive but the other is something I should've caught.                                                            │
&anonimal | Of those remaining 2, neither are in any critical areas.
```

The remaining 37 range across the entire base, from medium to high impact. Most are easily fixed. Some are rooted in Romako Shelkin design flaws and will require larger refactoring work.

Two of the SSU defects [are known](https://github.com/monero-project/kovri/issues/191) and should be addressed in #140. ping @EinMByte.

Regarding time-frame to complete: all of these would go away quicker if I was paid. I've never been paid to do any work on this project so the defects will eventually be chipped away with time. If anyone wants to speed up the process, send me an email and I'll send you an invite to view them.

**Comment by rakhimov** [comment]
Looking through some Coverity reported issues,
I noticed there is a fare amount of low level C style code,
for example, manual management of buffers, type-unsafe casts, C library functions.

Is it necessary to stay low-level for performance/portability/interface/hardware or other reasons for Kovri?

Would you mind if the code were refactored to be C++ instead?

**Comment by anonimal** [comment]
@rakhimov thank you for noticing this as this has been our biggest complaint from the **very beginning**. In our [earliest meetings](https://github.com/monero-project/kovri/issues/47), we agreed to move away from all C-style code unless it was absolutely necessary. Also, in #158, when I said:

>Not only does some of this code-base not even come close to complying with C++11 standards

I really meant it.

So, yes, @rakhimov, please, feel free to refactor! :sun_with_face:

**Comment by anonimal** [comment]
@EinMByte I think CID 151163 is a blaring false-positive, can you confirm? `SSUDataPacket` doesn't have a pointer to it's upcast and nowhere is `SSUDataPacket` (or any derived SSU class) ever created/deleted as a pointer to `SSUPacket` base.

**Comment by anonimal** [comment]
All that's left is a medium impact CID 151155 (which could be another false-positive) and low impact trivial move assignment optimizations in CID 143597/8 (which will be addressed in #366).

Thank you @EinMByte @rakhimov @guzzijones for helping to resolve #263 :+1:  :beers:

We can close this issue now and deal with future coverity issues as they come.
